src/libsystemd/sd-bus/sd-bus.c

```diff
@@ -2120,12 +2120,17 @@ int bus_ensure_running(sd_bus *bus) {
 
         assert(bus);
 
-        if (IN_SET(bus->state, BUS_UNSET, BUS_CLOSED, BUS_CLOSING))
-                return -ENOTCONN;
         if (bus->state == BUS_RUNNING)
                 return 1;
 
+        /* sd_bus_process() may call user callbacks which may unref the bus.
+         * Make sure we still have a valid ref. */
+        BUS_DONT_DESTROY(bus);
+
         for (;;) {
+                if (IN_SET(bus->state, BUS_UNSET, BUS_CLOSED, BUS_CLOSING))
+                        return -ENOTCONN;
```

Comment:
this is the *real* fix I think: so far we ended doing callouts to the Disconnected matches from bus_ensure_running() and we really shouldn't do that. bus_ensure_running() should never do callouts in fact, and I think the fact that it does is the bug. This change of yours should fix this however: once we notice that the connection is going down we will now fail instantly with ENOTOCONN instead of calling any callbacks.